### PR TITLE
Adds in the Networking layer API Call for bulk updating variations

### PR DESCRIPTION
**Pull request by chrisKns**
Part: #6240 

### Description
It will add a new [API call](https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-product-variations) in the Networking layer for executing bulk updates on an array of product variations.

### Testing instructions
No special test needed. Just running the unit tests.

### Screenshots
Non needed

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Mapper/ProductVariationMapper.swift [hunk 1]
```diff
@@ -28,6 +28,33 @@ struct ProductVariationMapper: Mapper {
     }
 }
 
+/// Mapper: ProductVariations
+///
+struct ProductVariationsMapper: Mapper {
```
Comment (Ecarrion) [inline Networking/Networking/Mapper/ProductVariationMapper.swift]: I think this name is a little bit confusing, especially because we already have a `ProductVariationListMapper`.
What do you think about a `ProductVariationsBulkUpdateMapper`, and I think it can exist in its own file too.

File: Networking/Networking/Mapper/ProductVariationMapper.swift [hunk 2]
```diff
@@ -41,3 +68,27 @@ private struct ProductVariationEnvelope: Decodable {
...
+///
+private struct ProductVariationsEnvelope: Decodable {
+    let updatedProductVariations: [ProductVariation]
+
+    private enum CodingKeys: String, CodingKey {
+        case update = "update"
+    }
+
+    private enum TopLevelCodingKeys: String, CodingKey {
+        case data = "data"
```
Comment (Ecarrion) [inline Networking/Networking/Mapper/ProductVariationMapper.swift]: nit: I think you can omit the explicit value.

File: Networking/Networking/Model/Product/ProductVariation.swift [hunk 3]
```diff
@@ -277,6 +277,9 @@ public struct ProductVariation: Codable, GeneratedCopiable, Equatable, Generated
 
         try container.encode(image, forKey: .image)
 
+        // Needed for bulk update variations, in other API calls it is ingored
+        try container.encode(productVariationID, forKey: .productVariationID)
```
Comment (Ecarrion) [inline Networking/Networking/Model/Product/ProductVariation.swift:281]: Could you explain why is this needed now?

File: Networking/Networking/Remote/ProductVariationsRemote.swift [hunk 4]
```diff
@@ -114,6 +115,31 @@ public class ProductVariationsRemote: Remote, ProductVariationsRemoteProtocol {
         }
     }
 
+    /// Updates a specific `ProductVariations`.
```
Comment (Ecarrion) [inline Networking/Networking/Remote/ProductVariationsRemote.swift]:
```suggestion
    /// Updates the provided `ProductVariations`.
```

File: Networking/Networking/Remote/ProductVariationsRemote.swift [hunk 5]
```diff
@@ -114,6 +115,31 @@ public class ProductVariationsRemote: Remote, ProductVariationsRemoteProtocol {
...
+    ///
+    /// - Parameters:
+    ///     - productVariations: the ProductVariations to update remotely.
+    ///     - completion: Closure to be executed upon completion.
+    ///
+    public func updateProductVariations(productVariations: [ProductVariation], completion: @escaping (Result<[ProductVariation], Error>) -> Void) {
+        guard let siteID = productVariations.first?.siteID, let productID = productVariations.first?.productID else {
```
Comment (Ecarrion) [inline Networking/Networking/Remote/ProductVariationsRemote.swift]: I think it is more clear, explicit, and safe to pass the `siteID` and `productID` as function parameterers.

File: Networking/Networking/Remote/ProductVariationsRemote.swift [hunk 6]
```diff
@@ -114,6 +115,31 @@ public class ProductVariationsRemote: Remote, ProductVariationsRemoteProtocol {
...
+            // If there an empty array is provided to update then simply return the same
+            completion(.success([]))
+            return
+        }
+
+        do {
+            let parameters = try productVariations.map({ try $0.toDictionary() })
```
Comment (Ecarrion) [inline Networking/Networking/Remote/ProductVariationsRemote.swift]:
```suggestion
            let parameters = try productVariations.map { try $0.toDictionary() }
```

**Review by Ecarrion — CHANGES_REQUESTED** [review]
Left some suggestions to make these functions a little bit more clear. Let me know what you think!

File: Networking/Networking/Model/Product/ProductVariation.swift (hunk 3 above)
Comment (chrisKns) [reply Networking/Networking/Model/Product/ProductVariation.swift:281]: The API Call for the bulk update sends as payload an array of product variations. The server expects for each object in the array to have an ID (so that it knows that variation to update).
Example payload:

```
{
  "update": [
    {
      "id": 733,
      "regular_price": "10.00",
....
    }
  ]
}
```

For other API calls that do not need the variation ID, this extra value is send, but the server ignores it. An other strategy could be to add an other property , something like `bulkUpdateProductVariationID`, that would be encoded if not nil, and set it to the value productVariationID has before the bulk update API call. But I though it was adding complexity without offering much.

File: Networking/Networking/Remote/ProductVariationsRemote.swift (hunk 5 above)
Comment (chrisKns) [reply Networking/Networking/Remote/ProductVariationsRemote.swift]: I agree, it makes the API more robust, and less likely to be used in the wrong way. I fixed it!

File: Networking/Networking/Mapper/ProductVariationMapper.swift (hunk 2 above)
Comment (chrisKns) [reply Networking/Networking/Mapper/ProductVariationMapper.swift]: Yes, I updated this 👍

File: Networking/Networking/Mapper/ProductVariationMapper.swift (hunk 1 above)
Comment (chrisKns) [reply Networking/Networking/Mapper/ProductVariationMapper.swift]: Yes I agree also latter we could extend it to add bulk create and bulk delete, I updated the PR 👍

File: Networking/Networking/Mapper/ProductVariationsBulkUpdateMapper.swift [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+///
+/// `Variations/batch` endpoint returns the requested update product variations document in a `update` key, nested in a `data` key.
+/// This entity allows us to do parse all the things with JSONDecoder.
+///
+private struct ProductVariationsEnvelope: Decodable {
+    let updatedProductVariations: [ProductVariation]
+
+    private enum CodingKeys: String, CodingKey {
+        case update = "update"
+        case data = "data"
```
Comment (Ecarrion) [inline Networking/Networking/Mapper/ProductVariationsBulkUpdateMapper.swift]:
```suggestion
        case update
        case data
```

File: Networking/Networking/Model/Product/ProductVariation.swift (hunk 3 above)
Comment (Ecarrion) [reply Networking/Networking/Model/Product/ProductVariation.swift:281]: I see now, this should be safe then. 
Thanks for the explanation!

**Review by Ecarrion — APPROVED** [review]
👍
